kube/src/client/mod.rs

```diff
@@ -130,35 +131,60 @@ impl Client {
-            Err(ws2::Error::Protocol(msg)) => Err(Error::RequestValidation(msg.into())),
-
-            // Fatal error with undelying connection.
-            Err(ws2::Error::Io(err)) => panic!("WebSocket connection error: {}", err),
-
-            // Unknown error. `tungstenite::Error` includes those that only happens after connecting.
-            Err(err) => panic!("Unknown WebSocket error: {}", err),
+            Err(err) => match err {
+                // tungstenite only gives us the status code.
+                ws2::Error::Http(code) => Err(Error::Api(ErrorResponse {
+                    status: code.to_string(),
+                    code: code.as_u16(),
+                    message: "".to_owned(),
+                    reason: "".to_owned(),
+                })),
+
+                ws2::Error::HttpFormat(err) => Err(Error::HttpError(err)),
+
+                // `tungstenite::Error::Tls` is only available when using `ws-native-tls` (`async-tungstenite/tokio-native-tls`)
+                // because it comes from `tungstenite/tls` feature.
+                #[cfg(feature = "ws-native-tls")]
+                ws2::Error::Tls(err) => Err(Error::SslError(format!("{}", err))),
+
+                // URL errors:
+                // - No host found in URL
+                // - Unsupported scheme (not ws/wss)
+                // shouldn't happen in our case
+                ws2::Error::Url(msg) => Err(Error::RequestValidation(msg.into())),
+
+                // Protocol errors:
+                // - Only GET is supported
+                // - Only HTTP version >= 1.1 is supported
+                // shouldn't happen in our case
+                ws2::Error::Protocol(msg) => Err(Error::RequestValidation(msg.into())),
+
+                ws2::Error::Io(err) => match err.kind() {
+                    std::io::ErrorKind::AddrNotAvailable
+                    | std::io::ErrorKind::ConnectionRefused
+                    | std::io::ErrorKind::ConnectionReset
+                    | std::io::ErrorKind::ConnectionAborted
+                    | std::io::ErrorKind::Interrupted
+                    | std::io::ErrorKind::InvalidInput
+                    | std::io::ErrorKind::InvalidData
+                    | std::io::ErrorKind::NotConnected
+                    | std::io::ErrorKind::NotFound
+                    | std::io::ErrorKind::TimedOut
+                    | std::io::ErrorKind::UnexpectedEof
+                    | std::io::ErrorKind::WouldBlock => Err(Error::Connection(err)),
+
+                    _ => panic!("WebSocket connection error: {}", err),
+                },
+
+                // Unexpected errors. `tungstenite::Error` contains errors that doesn't happen when trying to conect.
+                ws2::Error::ConnectionClosed
+                | ws2::Error::AlreadyClosed
+                | ws2::Error::Utf8
+                | ws2::Error::Capacity(_)
+                | ws2::Error::SendQueueFull(_) => {
+                    panic!("Unexpected {}", err);
```

Comment:
This is also probably not worthy of a `panic!`. If the call to `attach/exec` is part of a webapp, then sure, that part won't work, but we won't need to take down the entire app.

Reply:
These should never happen *when trying to connect* unless I'm misunderstanding. So there's something seriously wrong if it does.

- `ConnectionClosed` comes after `close` handshake which must be after the initial handshake (this connection) by definition.
- `AlreadyClosed` is even after that because it only happens if you tried to read/write on closed connection.
- `Utf8` is for receiving invalid text fame (I think)
- `Capacity` is about the message being too big
- `SendQueueFull` is about the message queue being full

If not panicking, should there be some generic error variant for those "totally unexpected, but probably not panic worthy" error?

Reply:
Ah, during connect. Yeah, I mean this is fine in the sense that we feel we "know it cannot happen", and we are explicit about it. However, it's not an __unrecoverable error__ for a larger application, which is what `panic!` is for.

If some part of a larger app tries to use kube ws features in some subroutine and it fails, it's fine to return an error type covering totally unexpected parts. Nothing bad will happen as a consequence (other than that part failing to work), it'll propagate up to whomever wants to handle it (or main).

Reply:
I guess the annoying thing is that we are splitting up `ws::Error` for user ergonomics and partitioning them into `kube::Error` types (which historically has just re-exported libraries' error types verbatim).

I think some kind of `WsOther` error which stringifies the remaining cases is probably good enough if we have a comment of the ones we expect in there.

Reply:
(I honestly think `tungstenite` should provide better error type that describes possible errors during connect, so that users like us don't need to guess and decide.)

I used `panic!` here because, to me, this is a broken contract and **is** unrecoverable. It's similar to when we use `.expect`. If the code reaches here, there _might_ be something truly wrong, and I'd rather have my larger app fail when this happens than to have an ambiguous error variant that is returned for unknown reason. No one knows what's safe anymore at this point.

> `panic!` is often appropriate if you’re calling external code that is out of your control and it returns an invalid state that you have no way of fixing.
>
> https://doc.rust-lang.org/book/ch09-03-to-panic-or-not-to-panic.html#guidelines-for-error-handling

> Nothing bad will happen as a consequence (other than that part failing to work)

I guess we just disagree here. I just have a more pessimistic view and can't say this :laughing:

I don't have a strong opinion. Especially because I doubt the code will ever reach here. If we're returning Err instead, what should I use?

Reply:
Sorry, I somehow didn't get your second comment until now. So my reply is just based on your first one.

Reply:
Added `WsOther`

Reply:
Hm.. Yeah, I understand the concern there. It would be a broken contract for the `connect` function to see, and it would probably cause further failures down the line for any continued calls to ws features (probably spamming continued errors in the logs of a larger application).

I can see a point for having that last `panic!` there. We would probably catch it early (since it's early stages of app functionality, after which it's possible nothing related to ws would work if it failed), it would indicate a bug in tungstenite, which could indicate a broken state, and maybe tungstenite is scribbling over memory as a result of that, but that feels like speculation to me. If it's possible however; that tungstenite can recover (through another connect say), so I am generally in favour of letting the app try (even if it's not great for it).

Still, this does feel like a bit of a philosophical point. Like in a normal k8s deployment setting, do we want:
1. degraded performance of pod(s) (where we potentially have lots of duplicate error logs, and non-working features within)
2. fully working pods, but with frequent crashes if something surfaces (potentially leading to full downtime if all replicas have frequent restarts on the same issue)

neither is great, and the choice is only for that time between it happening and when it's fixed.
2 is great when developing fast, but in a cloud with good observability or when working with inherited code (coding defensively); it's nice not causing outages on huge monoliths with http route or something. But that forces you to make the delineation of what's crucial enough to take down the app early if nothing panics (which usually leads to a period of spamming errors first).

Still, am on the fence about this particular one. panic is inflexible because it will cause downtime if we got it wrong (and propagated a panic in a legit recoverable case), but what they are doing IS wrong. If you are happy with that, I don't mind the panic really.